### ISSUE TITLE
new rule proposal for classes

### ISSUE THREAD

**Issue by Jxck** [eslint/eslint#4437]
I create plugin for classes.

https://github.com/Jxck/eslint-plugin-classes

works fine for me, and quite general so i'm suggest to merging this for default rules.
of course I can PR my code.

any comments are welcome.

detail from readme is below.

---

### space

should no space between method name and parens.

The following patterns are considered warnings:

```js
class Foo {
  bar () {
  }
}

class Foo {
  static bar () {
  }
}

class Foo {
  bar () {
  }

  buz() {
  }
}
```

The following patterns are not warnings:

```js
class Foo {
  bar() {
  }
}

class Foo {
  static bar() {
  }
}

class Foo {
  bar() {
  }

  buz() {
  }
}
```

### name

- class name should start with upper case.
- method name should start with lower case.

The following patterns are considered warnings:

```js
class foo {
}

class Foo {
  Bar () {
  }
}
```

The following patterns are not warnings:

```js
class Foo {
}

class Foo {
  bar () {
  }
}
```

## constructor

class should have constructor always even if empty body.

The following patterns are considered warnings:

```js
class Foo {
}

class Foo {
  bar() {
  }
}
```

The following patterns are not warnings:

```js
class Foo {
  constructor() {
    console.log('test');
  }
}

class Foo {
  constructor() {
  }

  bar() {
  }
}
```

## super

extended class should call `super()` at the top of constructor.

The following patterns are considered warnings:

```js
class Foo extends A {
  bar() {
  }
}

class Foo extends A {
  constructor() {
    this.bar();
  }
  bar() {
  }
}
```

The following patterns are not warnings:

```js
class Foo extends A {
  constructor() {
    super();
    this.a = 10;
  }
  bar() {
  }
}

class Foo extends A {
  bar() {
  }
  constructor() {
    super();
    this.a = 10;
  }
}

class Foo extends A {
  constructor() {
    // comments
    super();
    this.a = 10;
  }
}
```

## Style

enforce order of definition of method in order of static method => constructor => instance methods.

The following patterns are considered warnings:

```js
class Foo {
    constructor() {}
    static bar() {}
}


class Foo {
    baz() {}
    static bar() {}
    constructor() {}
    baz() {}
}
```

The following patterns are not warnings:

```js
class Foo {
    static bar() {}
    static bal() {}
    constructor() {}
    baz() {}
    biz() {}
    bez() {}
}
```

## Usage

```yaml
plugins:
  - classes

rules:
  # Plugins
  classes/space  : 2
  classes/name   : [2, "class", "method"]
  classes/constructor : 2
  classes/super  : 2
  classes/style  : 2
```

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by mysticatea** [comment]
Some rules seem to be existing in builtin, that `space` is the same as [space-before-function-paren](http://eslint.org/docs/rules/space-before-function-paren), and `super` is [constructor-super](http://eslint.org/docs/rules/constructor-super) and [no-this-before-super](http://eslint.org/docs/rules/no-this-before-super).

**Comment by Jxck** [comment]
@mysticatea  thanks for cover.
then rest proposals are

- classes/name
- classes/constructor
- classes/style

**Comment by ilyavolodin** [comment]
Classes name would already be partially covered by `new-cap`, not at the point of creation of the class, but at the point of instantiation. I'm not sure why `constructor` has to be present everywhere, but then I'm not use ES6, so it might be a valid rule. Classes style rule seems to be helpful, but I'm not sure how widely it would be used.

**Comment by nzakas** [comment]
I'd say for now let's leave these as custom rules. I think we already cover the most important cases, and you've created a plugin that covers the others. If we get requests for any of these in the future, we can revisit.
